File: api/db/supabase_client.py
"""
api/db/supabase_client.py
=========================
Supabase client wrapper for the Python backend (api, engine, ml layers).

- All credentials are loaded exclusively from environment variables.
- Exposes two singletons:
    supabase        → anon/public client  (safe for user-scoped operations)
    supabase_admin  → service-role client (server-side only, never expose to client)
- Import either singleton directly:
    from api.db.supabase_client import supabase, supabase_admin

Dependencies:
    pip install supabase python-dotenv
"""

import os
import logging
from functools import lru_cache

from dotenv import load_dotenv
from supabase import create_client, Client

from shared.constants import ConfigKey

# ---------------------------------------------------------------------------
# Bootstrap
# ---------------------------------------------------------------------------

load_dotenv()  # loads .env into os.environ (no-op in production where vars are injected)

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Config helpers
# ---------------------------------------------------------------------------

def _require_env(key: str) -> str:
    """Fetch a required env var; raise clearly if missing."""
    value = os.environ.get(key, "").strip()
    if not value:
        raise EnvironmentError(
            f"Missing required environment variable: '{key}'. "
            "Ensure it is set in your .env file or deployment environment."
        )
    return value


# ---------------------------------------------------------------------------
# Client factories (cached — one instance per process)
# ---------------------------------------------------------------------------

@lru_cache(maxsize=1)
def get_supabase_client() -> Client:
    """
    Returns the anon/public Supabase client.
    Safe for operations that respect Row Level Security (RLS).
    """
    url  = _require_env(ConfigKey.SUPABASE_URL)
    key  = _require_env(ConfigKey.SUPABASE_ANON_KEY)
    client = create_client(url, key)
    logger.info("Supabase anon client initialised (url=%s)", url)
    return client


@lru_cache(maxsize=1)
def get_supabase_admin_client() -> Client:
    """
    Returns the service-role Supabase client.
    Bypasses RLS — use ONLY for trusted server-side operations.
    Never expose this client or its key to the frontend.
    """
    url  = _require_env(ConfigKey.SUPABASE_URL)
    key  = _require_env(ConfigKey.SUPABASE_SERVICE_ROLE_KEY)
    client = create_client(url, key)
    logger.info("Supabase admin client initialised (url=%s)", url)
    return client


# ---------------------------------------------------------------------------
# Module-level singletons (convenience imports)
# ---------------------------------------------------------------------------

supabase: Client       = get_supabase_client()
supabase_admin: Client = get_supabase_admin_client()

File: extension/src/lib/supabaseClient.ts
/**
 * extension/src/lib/supabaseClient.ts
 * =====================================
 * Supabase client wrapper for the VS Code extension layer.
 *
 * - Credentials are loaded from VS Code extension settings / environment only.
 * - Exposes a single anon client — the extension NEVER holds the service-role key.
 * - The admin/service-role client lives exclusively in the Python backend.
 *
 * Dependencies:
 *   npm install @supabase/supabase-js
 *
 * Usage:
 *   import { getSupabaseClient } from './lib/supabaseClient';
 *   const client = getSupabaseClient();
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';
import * as vscode from 'vscode';
import { ConfigKey, ConfigDefault } from '../constants';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Shape of the Supabase config resolved at runtime. */
interface SupabaseConfig {
  url: string;
  anonKey: string;
}

// ---------------------------------------------------------------------------
// Config resolution
// ---------------------------------------------------------------------------

/**
 * Resolves Supabase credentials in priority order:
 *   1. VS Code workspace/user settings  (kaizencode.supabaseUrl / kaizencode.supabaseAnonKey)
 *   2. Process environment variables     (SUPABASE_URL / SUPABASE_ANON_KEY)
 *
 * Throws if either value is missing so failures are loud and early.
 */
function resolveConfig(): SupabaseConfig {
  const settings = vscode.workspace.getConfiguration('kaizencode');

  const url =
    settings.get<string>(ConfigKey.SUPABASE_URL) ||
    process.env['SUPABASE_URL'] ||
    '';

  const anonKey =
    settings.get<string>(ConfigKey.SUPABASE_ANON_KEY) ||
    process.env['SUPABASE_ANON_KEY'] ||
    '';

  if (!url) {
    throw new Error(
      `Supabase URL is not configured. ` +
      `Set '${ConfigKey.SUPABASE_URL}' in VS Code settings or the SUPABASE_URL env var.`
    );
  }

  if (!anonKey) {
    throw new Error(
      `Supabase anon key is not configured. ` +
      `Set '${ConfigKey.SUPABASE_ANON_KEY}' in VS Code settings or the SUPABASE_ANON_KEY env var.`
    );
  }

  return { url, anonKey };
}

// ---------------------------------------------------------------------------
// Singleton
// ---------------------------------------------------------------------------

let _client: SupabaseClient | null = null;

/**
 * Returns the shared Supabase anon client.
 * Initialised once on first call; subsequent calls return the cached instance.
 *
 * The client respects Row Level Security (RLS) — it operates as the
 * authenticated user or the anon role.
 */
export function getSupabaseClient(): SupabaseClient {
  if (!_client) {
    const { url, anonKey } = resolveConfig();
    _client = createClient(url, anonKey, {
      auth: {
        persistSession: true,
        autoRefreshToken: true,
        detectSessionInUrl: false, // not a browser — extension context
      },
    });
  }
  return _client;
}

/**
 * Resets the cached client (useful when credentials change at runtime,
 * e.g. user updates settings).
 */
export function resetSupabaseClient(): void {
  _client = null;
}
